app/Http/Controllers/Api/BookController.php

```diff
@@ -0,0 +1,107 @@
+
+            //Get all comments ID of posts was deleted.
+            $commentsID = Comment::whereIn('post_id', $postsID)->pluck('id')->toArray();
+
+            //Delete all comments and its favorites.
+            Comment::whereIn('post_id', $postsID)->update(['deleted_at' => $timeDelete, 'updated_at' => $timeDelete]);
+            Favorite::whereIn('favoritable_id', $commentsID)->where('favoritable_type', Favorite::TYPE_COMMENT)->update(['deleted_at' => $timeDelete, 'updated_at' => $timeDelete]);
+            DB::commit();
+        } catch (\PDOException $e) {
+            DB::rollBack();
+        }
+        if ($request->ajax()) {
+            return response()->json(['book'=> $book, 'time' => $timeDelete], 200);
+        }
+    }
+
+    /**
+     * Restore book and its relationship with id of book.
+     *
+     * @param Request $request request
+     * @param int     $id      id of book
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function restore(Request $request, $id)
+    {
+        $timeDelete = Book::withTrashed()->where('id', $id)->pluck('deleted_at')->first();
+        DB::beginTransaction();
+        try {
+            //Restore book and its favorite.
+            $book = Book::withTrashed()->find($id)->restore();
+            Favorite::withTrashed()->where('deleted_at', $timeDelete)->where('favoritable_type', Favorite::TYPE_BOOK)->where('favoritable_id', $id)->restore();
+
+            //Restore rating, qrcode, borrowing.
+            Rating::withTrashed()->where('book_id', $id)->where('deleted_at', $timeDelete)->restore();
+            QrCode::withTrashed()->where('book_id', $id)->where('deleted_at', $timeDelete)->restore();
+            Borrowing::withTrashed()->where('book_id', $id)->where('deleted_at', $timeDelete)->restore();
+
+            //Get all posts ID was restored.
+            $postsID = Post::withTrashed()->where('book_id', $id)->where('deleted_at', $timeDelete)->pluck('id')->toArray();
+
+            //Restore all post of this book and its favorite.
+            Post::withTrashed()->where('book_id', $id)->where('deleted_at', $timeDelete)->restore();
+            Favorite::withTrashed()->where('deleted_at', $timeDelete)->where('favoritable_type', Favorite::TYPE_POST)->whereIn('favoritable_id', $postsID)->restore();
+
+            //Get all comments ID of posts was restored.
+            $commentsID = Comment::withTrashed()->where('deleted_at', $timeDelete)->whereIn('post_id', $postsID)->pluck('id')->toArray();
+
+            //Restore all comments and its favorite.
+            Comment::withTrashed()->whereIn('post_id', $postsID)->where('deleted_at', $timeDelete)->restore();
+            Favorite::withTrashed()->where('deleted_at', $timeDelete)->where('favoritable_type', Favorite::TYPE_COMMENT)->whereIn('favoritable_id', $commentsID)->restore();
+            DB::commit();
+        } catch (\PDOException $e) {
```

Comment:
you should add catch Exception, because sometime will has exception which is not a PDOException.